common/rpc/rpc.go

```diff
@@ -162,19 +162,20 @@ func (d *RPCFactory) GetInternodeClientTlsConfig() (*tls.Config, error) {
 
 // GetGRPCListener returns cached dispatcher for gRPC inbound or creates one
 func (d *RPCFactory) GetGRPCListener() net.Listener {
-	d.initListener.Do(func() {
-		hostAddress := net.JoinHostPort(getListenIP(d.config, d.logger).String(), convert.IntToString(d.config.GRPCPort))
-		var err error
-		d.grpcListener, err = net.Listen("tcp", hostAddress)
+	return d.grpcListener()
+}
 
-		if err != nil {
-			d.logger.Fatal("Failed to start gRPC listener", tag.Error(err), tag.Service(d.serviceName), tag.Address(hostAddress))
-		}
+func (d *RPCFactory) createGRPCListener() net.Listener {
+	hostAddress := net.JoinHostPort(getListenIP(d.config, d.logger).String(), convert.IntToString(d.config.GRPCPort))
+	var err error
+	grpcListener, err := net.Listen("tcp", hostAddress)
 
-		d.logger.Info("Created gRPC listener", tag.Service(d.serviceName), tag.Address(hostAddress))
-	})
+	if err != nil {
+		d.logger.Fatal("Failed to start gRPC listener", tag.Error(err), tag.Service(d.serviceName), tag.Address(hostAddress))
+	}
 
-	return d.grpcListener
```

Comment:
@dnr I was thinking `d.grpcListener` may not synchronize between the goroutine that does the assignment in line 168 and another one which just waits for `Do` without running the func. In that case, the second goroutine may return nil. Is that right?

Reply:
I thought that was maybe possible, but if you look at the implementation of Once, there should always be synchronization there, either Load observes the Store, or the mutex Lock observes the Unlock.

Actually even if you don't look at implementation, anything that Once could do to implement its guarantee "no call to Do returns until the one call to f returns" must synchronize somehow.

And if you look at OnceValue, it's just wrapping this pattern.

Reply:
Makes sense.

Not sure how to explain the error though unless net.Listen returned `nil, nil`, or returned a listener with nil `Addr()`. I'm going to add those checks.